### DISCUSSION TITLE
Lesson 10: ERC20 AI test function error

### DISCUSSION THREAD

**Discussion by AbhiSharma1210** [Cyfrin/foundry-full-course-cu#1035] [Q&A]
So as per the lesson I asked AI to generate test script and following are the functions form the AI test script that I used.
.
.
.
`function testTransferToken() public {
        // Bob transfers tokens to Alice
        vm.prank(bob);
        uint256 transferAmount = 50;
        ourToken.transfer(alice, transferAmount);

        assertEq(ourToken.balanceOf(alice), transferAmount);
        assertEq(ourToken.balanceOf(bob), STARTING_BALANCE - transferAmount);
    }

    function testTransferWithInsufficientBalance() public {
        // Bob attempts to transfer more tokens than his balance
        vm.prank(bob);
        uint256 transferAmount = STARTING_BALANCE + 1;
        vm.expectRevert();

        ourToken.transfer(alice, transferAmount);

        assertEq(ourToken.balanceOf(alice), 0); // Alice's balance remains unchanged
        assertEq(ourToken.balanceOf(bob), STARTING_BALANCE); // Bob's balance remains unchanged
    }

    function testTransferToZeroAddress() public {
        // Bob attempts to transfer to the zero address
        vm.prank(bob);
        uint256 transferAmount = 10;
        bool success = ourToken.transfer(address(0), transferAmount);

        assertEq(ourToken.balanceOf(bob), STARTING_BALANCE); // Bob's balance remains unchanged
        assertFalse(success); // The transfer should fail
    }

    function testTransferFromWithInsufficientAllowance() public {
        // Alice attempts to transferFrom Bob with insufficient allowance
        vm.prank(alice);
        uint256 transferAmount = 10;
        bool success = ourToken.transferFrom(bob, alice, transferAmount);

        assertEq(ourToken.balanceOf(alice), 0); // Alice's balance remains unchanged
        assertEq(ourToken.balanceOf(bob), STARTING_BALANCE); // Bob's balance remains unchanged
        assertFalse(success); // The transfer should fail
    }`
Now all the functions here are using `bool success` and then `assertFalse` in the end and they are throwing error as follows 
for `testTransferToZeroAddress`
`[FAIL. Reason: ERC20InvalidReceiver(0x0000000000000000000000000000000000000000)] testTransferToZeroAddress() (gas: 10600)
Test result: FAILED.`
except `testTransferWithInsufficientBalance()` (I modified this due to error and it ran successfully).
My question is whether it is possible to successfully run the "Failed" AI function with the way its trying to do or should I just go with `vm.expectRevert` only. I have also gone through the 'assertFalse' documentation on foundry and the explanation is not very detailed.

**Comment by theirrationalone** [comment]
Yes it's possible with a specific function naming convention......

You can command AI to name test functions which you intent to `Fail` like this below....

E.g....

```solidity
function testFailsFunction() public {
        uint256 a = 1;
        uint256 b = 2;
        assertEq(a, b);
}

// or

function testFailedFunction() public {
        uint256 a = 1;
        uint256 b = 2;
        assertEq(a, b);
}
```

My prompt will be:
## Hey AI, Please test this SMC and Please name all failing test functions prepended with these specific chords `testFailed`,  `testFails`.
For Example: If i have to test unequal balances of two persons then that test function name would be like this... `function testFailsBalanceCheck() public {}`.

**Reply by AbhiSharma1210** [reply]
My another question is why is console throwing error when `bool success` and `assertFalse` are used. syntactically this looks fine to me. 
for example this function.

```
function testTransferFromWithInsufficientAllowance() public {
    // Alice attempts to transferFrom Bob with insufficient allowance
    vm.prank(alice);
    uint256 transferAmount = 10;
    bool success = ourToken.transferFrom(bob, alice, transferAmount);
    assertEq(ourToken.balanceOf(alice), 0); // Alice's balance remains unchanged
    assertEq(ourToken.balanceOf(bob), STARTING_BALANCE); // Bob's balance remains unchanged
    assertFalse(success); // The transfer should fail
}
```

**Reply by theirrationalone** [reply]
Ohhh Yes,

It's an AI generated Test and AI might not have any info about `ERC20 token`...

look at this chunk of `transferFrom` function from ERC20 contract, below....
```solidity
function transferFrom(
        address from,
        address to,
        uint256 amount
    ) public virtual override returns (bool) {
        address spender = _msgSender();
        _spendAllowance(from, spender, amount); // could throw an error on invalid or insufficient allowance. This is the issue in your case.
        _transfer(from, to, amount);
        return true; // only true is returned on a successful transfer.
}
```

You can see there is no `bool false` returned.... We could have either a `bool true` on success or aa `error` on (invalid or insufficient allowance).

Your test function is crashing before coming to `assertFalse`...

```solidity
function testTransferFromWithInsufficientAllowance() public {
    // Alice attempts to transferFrom Bob with insufficient allowance
    vm.prank(alice); 
    uint256 transferAmount = 10;
    // here we didn't approved alice intentionally. 😉 read this comment 👈
    bool success = ourToken.transferFrom(bob, alice, transferAmount); // an error on `insufficient allowance` thrown. 😉 read this comment 👈
   // test ended here above due to a crash (error exception not handled). 😉 read this comment 👈
    assertEq(ourToken.balanceOf(alice), 0); // Alice's balance remains unchanged
    assertEq(ourToken.balanceOf(bob), STARTING_BALANCE); // Bob's balance remains unchanged
    assertFalse(success); // The transfer should fail
}
```

You just need learn Documentations (comments) embedded with `ERC20s`. You can paste dependencies in your AI prompt to give directions to use them within before actual tests.

**Reply by AbhiSharma1210** [reply]
Thank you for the explanation but I knew why the error is occurring in the AI generated tests that's why I also mentioned that I refactored one of the functions (mentioned below) using vm.expectRevert and that worked. So in my point of view this approach is better. I was curious if its possible to successfully run the test with the AI approach but it seems that its may be very hard to achieve.

```solidity
function testTransferWithInsufficientBalance() public {
    // Bob attempts to transfer more tokens than his balance
    vm.prank(bob);
    uint256 transferAmount = STARTING_BALANCE + 1;
    vm.expectRevert();

    ourToken.transfer(alice, transferAmount);

    assertEq(ourToken.balanceOf(alice), 0); // Alice's balance remains unchanged
    assertEq(ourToken.balanceOf(bob), STARTING_BALANCE); // Bob's balance remains unchanged
}
```

**Reply by theirrationalone** [reply]
Yeahh, requires sophisticated contextual prompting 🧲
